### PR TITLE
JIT: initial workup for suppressing dumps and checks post-phase

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Allow phases to declare what parts of jit state they might have modified,
and suppress dumps and checks based on these declarations.

Existing phases that don't know about this are handled by defaulting to
reporting that everything might have beem nodified.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib curious what you think of this approach. If it looks reasonable I will propagate to more phases.

Might be overkill to return fine-grained `PhaseStatus`; it could just be a bool.

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib any thoughts on this proposed change to phases?

File: src/coreclr/src/jit/compiler.h [hunk 1]
```diff
@@ -1130,6 +1130,17 @@ enum class PhaseChecks
     CHECK_ALL
 };
 
+// Specify compiler data that a phase might modify
+enum class PhaseStatus : unsigned
+{
+    PS_MODIFIED_NOTHING    = 0x0,
```
Comment (sandreenko) [inline src/coreclr/src/jit/compiler.h]: Nit: enum class should not need a class prefix, because they can't be used without className.

**Review by sandreenko — APPROVED** [review]
I like the idea in general, I am not sure how much will it affect log sizes and checked TP. 
Probably nobody is scrolling through the whole JitDump, but if you are looking for a specific tree and you will get twice less occurrences it will speed up the process significantly.

File: src/coreclr/src/jit/rationalize.cpp [hunk 2]
```diff
@@ -817,7 +817,16 @@ Compiler::fgWalkResult Rationalizer::RewriteNode(GenTree** useEdge, Compiler::Ge
...
 
-void Rationalizer::DoPhase()
+//------------------------------------------------------------------------
+// DoPhase: Run the rationalize over the method IR.
+//
+// Returns:
+//    PhaseStatus indicating, what, if anything, was modified
+//
+// Notes:
+//    Runs only if Compiler::optMethodFlags has flag OMF_HAS_NEWOBJ set.
```
Comment (CarolEidt) [inline src/coreclr/src/jit/rationalize.cpp]: Copy/paste error?

File: src/coreclr/src/jit/stacklevelsetter.cpp [hunk 3]
```diff
@@ -55,6 +59,9 @@ void StackLevelSetter::DoPhase()
 
     comp->fgSetPtrArgCntMax(maxStackLevel);
     CheckArgCnt();
+
+    // Might want an "other" category for things like this...
```
Comment (CarolEidt) [inline src/coreclr/src/jit/stacklevelsetter.cpp]: Any idea what checking such an "other" category might want to trigger?

**Review by CarolEidt — COMMENTED** [review]
Looks like a good approach to me.

**Review by BruceForstall — APPROVED** [review]
LGTM.

Seems semi-dangerous to trust phases to report what changed and then not check. Seems ok to not dump in that case.

File: src/coreclr/src/jit/compiler.h (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/compiler.h]: Good point, it is annoying to repeat this.

File: src/coreclr/src/jit/stacklevelsetter.cpp (hunk 3 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/stacklevelsetter.cpp]: Not sure? Seems like a good part of the compiler object state probably can't be checked effectively. But maybe there are some things we can come up with.

File: src/coreclr/src/jit/rationalize.cpp (hunk 2 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/rationalize.cpp]: Yes.

**Comment by AndyAyersMS** [comment]
Need to merge in latest...

**Comment by AndyAyersMS** [comment]
Moving out of draft status.

@dotnet/jit-contrib any other thoughts?

**Review by CarolEidt — APPROVED** [review]
LGTM

File: src/coreclr/src/jit/flowgraph.cpp [hunk 4]
```diff
@@ -24151,16 +24180,18 @@ void Compiler::fgRemoveEmptyFinally()
             printf("\n");
         }
```
Comment (briansull) [inline src/coreclr/src/jit/flowgraph.cpp]: GitHub bug: All the real changes are not shown

File: src/coreclr/src/jit/phase.cpp [hunk 5]
```diff
@@ -88,45 +87,60 @@ void Phase::PrePhase()
...
+//
+void Phase::PostPhase(PhaseStatus status)
 {
 #ifdef DEBUG
 
+    // Don't dump or check post phase unless the phase made changes.
+    const bool        madeChanges   = (status != PhaseStatus::MODIFIED_NOTHING);
+    const char* const statusMessage = madeChanges ? "" : " [no changes]";
+    bool              doPostPhase   = false;
+
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/coreclr/src/jit/phase.cpp]: Good idea. I'll implement this, and also check that some other easy to verify things are unchanged.

**Review by briansull — APPROVED** [review]
LGTM, Thanks

**Comment by AndyAyersMS** [comment]
Am going to ignore the OSX libraries test failure, suspect this is when we briefly lost connectivity to the test machines.

**Comment by AndyAyersMS** [comment]
Part of #2109.

**Pull request merged**